### ISSUE TITLE
Certificate use in Code flow?

### ISSUE THREAD

**Issue by marko-grujic-frame** [OneDrive/onedrive-api-docs#962]
Hi there,

reading through the [documentation](https://docs.microsoft.com/en-us/onedrive/developer/rest-api/getting-started/graph-oauth?view=odsp-graph-online#step-2-redeem-the-code-for-access-tokens) on authentication for the OneDrive API using the code flow it remains unclear whether a certificate can be used as a client secret to redeem the code instead of a password?

Thanks,
Marko

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 210daf87-f00d-db74-cdb0-279d033d29de
* Version Independent ID: 03050364-41d2-e475-69b1-08b554ade330
* Content: [Authorization for OneDrive API via Microsoft Graph - OneDrive dev center](https://docs.microsoft.com/en-us/onedrive/developer/rest-api/getting-started/graph-oauth?view=odsp-graph-online#feedback)
* Content Source: [docs/rest-api/getting-started/graph-oauth.md](https://github.com/OneDrive/onedrive-api-docs/blob/live/docs/rest-api/getting-started/graph-oauth.md)
* Product: **onedrive**
* GitHub Login: @rgregg
* Microsoft Alias: **rgregg**

**Comment by marko-grujic-frame** [comment]
So through some trial-and-error I managed to get this working I think (I employed a hybrid approach taking clues from [another azure oauth flow](https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-on-behalf-of-flow#second-case-access-token-request-with-a-certificate)).

In short, I just replaced the `client_secret` parameter with:

1. `client_assertion_type` set to `urn:ietf:params:oauth:client-assertion-type:jwt-bearer`
2. `client_assertion` set to a JWT token

The JWT token was generated according to [this recipe](https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-certificate-credentials), with one important exception: the `x5t` header was not simply a "X.509 Certificate SHA-1 thumbprint", but instead a transformed version as I discovered from the ADAL library's `_create_x5t_value` method. In plain python:
```python
hex_val = binascii.a2b_hex(thumbprint) 
x5t = base64.urlsafe_b64encode(hex_val).decode()
```

Prior to closing the issue I would appreciate someone from MS acknowledging whether this is a proper way to handle this case.

**Comment by marko-grujic-frame** [comment]
So through some trial-and-error I managed to get this working I think (I employed a hybrid approach taking clues from [another azure oauth flow](https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-on-behalf-of-flow#second-case-access-token-request-with-a-certificate)).

In short, I just replaced the `client_secret` parameter with:

1. `client_assertion_type` set to `urn:ietf:params:oauth:client-assertion-type:jwt-bearer`
2. `client_assertion` set to a JWT token

The JWT token was generated according to [this recipe](https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-certificate-credentials), with one important exception: the `x5t` header was not simply a "X.509 Certificate SHA-1 thumbprint", but instead a transformed version as I discovered from the ADAL library's `_create_x5t_value` method. In plain python:
```python
hex_val = binascii.a2b_hex(thumbprint) 
x5t = base64.urlsafe_b64encode(hex_val).decode()
```

Prior to closing the issue I would appreciate someone from MS acknowledging whether this is a proper way to handle this case.
